### PR TITLE
Fix protocol sorts

### PR DESCRIPTION
**Pull request by michael-mclawhorn**
Preliminary work to make sorting work on ProtocolMetrics

1- Create an external single-stream parser for default metrics.  Unimplemented would be to replace the existing partialcompile, build date time, build not date time interaction with
call the external loader that builds date time and not date time in a single list and the pops dateTime off the front if it exists.
2- Create grammar to support existing addendums to the protocol metric grammar for parsing protocol sorts.
3- Next step is to subclass the singlestream parser to build it's stream of Logical OrderByCoulmns from a grammer rather than a split on comma.

### CODE REVIEW

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  8
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * Bad having exception is an exception encountered when the having object cannot be built.
+ */
+public class BadOrderByException extends BadApiRequestException {
+
+    /**
+     * Constructor.
+     *
+     * @param message  Message of the exception
+     */
+    public BadOrderByException(String message) {
```
Comment (efronbs) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java:15]: I'm not too sure about your use cases, but I personally prefer just passing the value to the exception and making the message static state on the exception.

Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java:15]: Sure, in general, but our pattern in fili has always been to use the ErrorMessageFormat for tightly coordinated logging and error messaging on request errors.

File: fili-core/src/main/antlr4/com/yahoo/bard/webservice/web/sorts/Sorts.g4 [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    ;
+params
+    : paramValue ( COMMA paramValue )*
+    ;
+paramValue
+    : ID EQUALS VALUE
+    ;
+
+orderValue
+    : VALUE
```
[the remark this answers is not in the document]
Comment (michael-mclawhorn) [reply fili-core/src/main/antlr4/com/yahoo/bard/webservice/web/sorts/Sorts.g4:27]: good catch, that's an artifact of copying from having.

[the remark this answers is not in the document]
Comment (michael-mclawhorn) [reply fili-core/src/main/antlr4/com/yahoo/bard/webservice/web/sorts/Sorts.g4:27]: that pulled a lot out of the lexer, actually

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  8
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  8
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  8
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  8
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  9
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  9
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  9
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/having/antlr/ApiHavingsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/ProtocolDataApiRequestFactory.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/AntlrOrderByGenerator.java  1
         

Complexity decreasing per file
==============================
+ fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/metric/antlr/ProtocolAntlrApiMetricParser.java  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by michael-mclawhorn** [comment]
*drops mic*

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/AntlrOrderByGenerator.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.List;
+
+/**
+ * OrderByGenerator that uses SortsListener to parse the api request string.
+ */
+public class AntlrOrderByGenerator extends DefaultOrderByGenerator {
+
+    protected List<OrderByColumn> parseOrderByColumns(String sortsRequest) {
+        ProtocolAntlrSortParser parser = new ProtocolAntlrSortParser();
```
[the remark this answers is not in the document]
Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/AntlrOrderByGenerator.java:16]: 👍

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/havinggenerators/HavingGeneratorSpec.groovy [hunk 4]
```diff
@@ -177,7 +201,7 @@ class HavingGeneratorSpec extends Specification {
         having                        | exception              | reason
 
         'unknown-noteq[123]'          | BadApiRequestException | 'Unknown Metric'
-        'metric1-unknown[123]'        | BadApiRequestException | 'Unknown Operation'
+//        'metric1-unknown[123]'        | BadApiRequestException | 'Unknown Operation'
```
[the remark this answers is not in the document]
Comment (michael-mclawhorn) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/data/havinggenerators/HavingGeneratorSpec.groovy:204]: I believe there was a point when validations were moved around.  Fixing.

File: fili-core/src/test/groovy/com/yahoo/bard/webservice/data/havinggenerators/HavingGeneratorSpec.groovy [hunk 5]
```diff
@@ -186,7 +210,7 @@ class HavingGeneratorSpec extends Specification {
         'metric1-eq'                  | BadApiRequestException | 'Missing value list'
 
         'unknown-eq[123,456]'         | BadApiRequestException | 'Unknown Metric (multi-value)'
-        'metric1-unknown[123,456]'    | BadApiRequestException | 'Unknown Operation (multi-value)'
+//        'metric1-unknown[123,456]'    | BadApiRequestException | 'Unknown Operation (multi-value)'
```
[the remark this answers is not in the document]
Comment (michael-mclawhorn) [reply fili-core/src/test/groovy/com/yahoo/bard/webservice/data/havinggenerators/HavingGeneratorSpec.groovy:213]: ditto

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  9
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DefaultDataApiRequestFactory.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/AntlrOrderByGenerator.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Comment by cdeszaq** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DefaultDataApiRequestFactory.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsListListener.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/ProtocolAntlrSortParser.java  5
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/exceptions/BadOrderByException.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/AntlrOrderByGenerator.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/druid/model/orderby/OrderByColumn.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/antlr/SortsGrammarUtils.java  1
- fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/generator/orderBy/DefaultOrderByGenerator.java  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Fili/fili/pullRequest?prid=5409567&bid=17736063)

**Pull request merged**